python/lsst/daf/butler/registry/bridge/monolithic.py

```diff
@@ -185,24 +184,54 @@ def check(self, refs: Iterable[DatasetIdRef]) -> Iterable[DatasetIdRef]:
             yield byId[row["dataset_id"]]
 
     @contextmanager
-    def emptyTrash(self) -> Iterator[Iterable[DatasetIdRef]]:
+    def emptyTrash(self, records_table: Optional[Any] = None,
+                   record_class: Optional[Type[StoredDatastoreItemInfo]] = None
+                   ) -> Iterator[Iterable[Tuple[DatasetIdRef, Optional[StoredDatastoreItemInfo]]]]:
         # Docstring inherited from DatastoreRegistryBridge
-        sql = sqlalchemy.sql.select(
-            [self._tables.dataset_location_trash.columns.dataset_id]
+
+        if records_table is None:
+            raise ValueError("This implementation requires a records table.")
+
+        if record_class is None:
+            raise ValueError("Record class must be provided if records table is given.")
+
+        # SELECT records.dataset_id, records.path FROM records
+        #    JOIN records on dataset_location.dataset_id == records.dataset_id
+        #    WHERE dataset_location.datastore_name = datastoreName
+
+        # It's possible that we may end up with a ref listed in the trash
+        # table that is not listed in the records table. Such an
+        # inconsistency would be missed by this query.
+        sql = records_table._table.select(
         ).select_from(
             self._tables.dataset_location_trash
         ).where(
-            self._tables.dataset_location_trash.columns.datastore_name == self.datastoreName
+            sqlalchemy.sql.and_(
+                self._tables.dataset_location_trash.columns.dataset_id
+                == records_table._table.columns.dataset_id,
+                self._tables.dataset_location_trash.columns.datastore_name == self.datastoreName
```

Comment:
Took me a while to figure out the right SQL (and that's easier that trying to explain it in words), but this is what I had in mind for a way to remove the per-dataset `_can_remove_dataset_artifact` call:
```suggestion
                self._tables.dataset_location_trash.columns.datastore_name == self.datastoreName,
                sqlalchemy.sql.not_(
                    path_column.in_(
                        # Subquery for all paths in dataset_locations (not trash),
                        # as we don't want to delete any artifact still in use
                        # there.
                        sqlalchemy.sql.select(
                            [path_column]
                        ).select_from(
                            self._tables.dataset_location.join(
                            records_table._table,
                            onclause=(
                                self._tables.dataset_location.columns.dataset_id ==
                                records_table._table.columns.dataset_id
                            ),
                        ).where(
                            self._tables.dataset_location_trash.columns.datastore_name == self.datastoreName
                        )
                    )
                )
```
That requires
```
path_column = records_table._table.columns["path"]
```
to be defined earlier.  But it's kind of encapsulation-breaking for the generic bridge manager to know that the "path" column is special or even that a query over something like paths is relevant for this datastore, and I don't know how best to deal with that.

Reply:
Yes, up to now all we are saying is that there is an id in this table matching an id in that other one and when you do the join you can throw all the content at this class I'm giving you.  It does seem a bit of a step to jump from there to requiring that a path column exists. I can try your code and see how fast it makes it and compare it with doing a separate IN query. I guess it doesn't help that the `path` is a TEXT field and not part of an index.

Reply:
Also, now that I'm trying this, we still need to delete the rows from records and trash even if the file is owned by multiple refs. This is controlling whether to delete the artifact, not whether to delete the ref from datastore. Won't any combo query lose the rows that must be deleted even if the artifacts stay unless the resulting row has a flag in it saying whether to keep the artifact or delete it.

Reply:
We could rewrite the query to make the "can_delete_artifact" a returned `bool` column, but that'd complicate constructing `StoredItemInfo` from the result row.

How about we make this two queries instead of one (or N+1):
```
# Helper closure to generate the common join+where clause.
def join_records(select: sqlalchemy.sql.Select, location_table: sqlalchemy.schema.Table) -> sqlalchemy.sql.FromClause:
    return select.select_from(
        record_table._table.join(
            location_table,
            onclause=record_table._table.columns.dataset_id == location_table.columns.dataset_id,
        )
    ).where(
        location_table.columns.datastore_name == self.datastoreName
    )

# Tim's original query (`sql`)
info_in_trash = join_records(record_table._table.select(), self._tables.dataset_location_trash)

# Some helper subqueries
paths_not_in_trash = join_records(
    sqlalchemy.sql.select([record_table._table.columns["path"]]),
    self._tables.dataset_location,
).alias("paths_not_in_trash")
paths_in_trash = join_records(
    sqlalchemy.sql.select([record_table._table.columns["path"]]),
    self._tables.dataset_location_trash,
).alias("paths_in_trash")

# A query for paths that are referenced by datasets in the trash and datasets not in the trash.
paths_to_preserve = sqlalchemy.sql.select(
    [paths_in_trash.columns["path"]]
).select_from(
    paths_not_in_trash.join(
        paths_in_trash,
        onclause=paths_in_trash.columns["path"] == paths_not_in_trash.columns["path"]
    )
)
```
I'm thinking the bridge manager would execute both queries and return via the context manager both the list of `StoredItemInfo` and a `set` of paths that need to be preserved.  The string column name `"path"` would be another argument passed by the Datastore; if it's `None`, we just wouldn't do the second query.

And I'm less uncomfortable than I was yesterday with having the bridge class run this query for the datastore - it's essentially asking the datastore for the name of a column that uniquely identifies its artifacts and offering to find artifacts shared across the trash boundary, which seems reasonably generic.

And yes, I think we probably do want an index on `path` in the future, when we can do it as a (trivial) migration.  For now, I think we'll still get a huge speedup without it, because I'm 95% sure it's the latency of a per-dataset query that's killing us, and just moving the repeated no-index searches into a loop inside the database will be enough of a speedup for now.

Reply:
I did the `in` clause query outside of the bridge manager and timing went from 8 seconds to 5 seconds (`master` takes 13). The actual timing for deleting 1000 datasets (with real timing in the emptyTrash method) is something like.

Before:
```
INFO  2021-05-07T12:07:35.102-0700 lsst.daf.butler.datastores.fileDatastore ()(fileDatastore.py:1650)- Emptying trash
INFO  2021-05-07T12:07:35.106-0700 lsst.daf.butler.datastores.fileDatastore ()(fileDatastore.py:1653)- Requested trashed datasets
INFO  2021-05-07T12:07:42.471-0700 lsst.daf.butler.datastores.fileDatastore ()(fileDatastore.py:1706)- Trash complete
```

with join as on this ticket and in operator to create set of all refs per file:

```
INFO  2021-05-07T12:07:59.412-0700 lsst.daf.butler.datastores.fileDatastore ()(fileDatastore.py:1633)- Emptying trash
INFO  2021-05-07T12:07:59.440-0700 lsst.daf.butler.datastores.fileDatastore ()(fileDatastore.py:1649)- Done with database
INFO  2021-05-07T12:08:00.721-0700 lsst.daf.butler.datastores.fileDatastore ()(fileDatastore.py:1687)- Trash complete
```
so without all the command line overhead we have the whole thing complete now in just over a second whereas before it takes 7 seconds.

This seems good. I think doing it your way above probably leads to a lower memory overhead and less risk of IN operator going bonkers with a million files.

Reply:
For completeness, the table querying overhead in this version of the code is 3 seconds (so half the time for half the number of select calls).